mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp

```diff
@@ -772,6 +787,964 @@ struct GridwiseGemmRewritePattern : public OpRewritePattern<GridwiseGemmOp> {
+    const int64_t waveSize = rock::lookupArchInfo(arch).waveSize;
+
+    TypedValue<MemRefType> inQ = op.getQueries();
+    ArrayRef<int64_t> qShape = inQ.getType().getShape();
+    Type elemTypeQ = inQ.getType().getElementType();
+
+    TypedValue<MemRefType> inK = op.getKeys();
+    ArrayRef<int64_t> kShape = op.getKeys().getType().getShape();
+    Type elemTypeK = inK.getType().getElementType();
+
+    TypedValue<MemRefType> inV = op.getValues();
+    Type elemTypeV = inV.getType().getElementType();
+
+    TypedValue<MemRefType> out = op.getOut();
+    ArrayRef<int64_t> outShape = op.getOut().getType().getShape();
+    Type elemTypeOut = out.getType().getElementType();
+
+    // Gemm0 out is casted to be elemTypeV
+    Type elemTypeQxK = elemTypeV;
+
+    auto privateMemoryAddressSpace = rewriter.getAttr<gpu::AddressSpaceAttr>(
+        gpu::GPUDialect::getPrivateAddressSpace());
+
+    int64_t gemm0G = qShape[0];
+    int64_t gemm0K = qShape[1];
+    int64_t gemm0M = qShape[2];
+    int64_t gemm0N = kShape[2];
+
+    int64_t gemm1M = outShape[1];
+    int64_t gemm1N = outShape[2];
+
+    RockAccelTuningParamAttrInterface gemm0TuningParams = op.getParams();
+    int64_t gemm0kpack = gemm0TuningParams.getKpack();
+    int64_t gemm0KpacksPerBlock = gemm0TuningParams.getKpackPerBlock();
+    int64_t gemm0MPerBlock = gemm0TuningParams.getMPerBlock();
+    int64_t gemm0NPerBlock = gemm0TuningParams.getNPerBlock();
+    bool forceUnroll = gemm0TuningParams.getForceUnroll();
+    int64_t gemm0MBlocks = gemm0M / gemm0MPerBlock;
+    int64_t gemm0NBlocks = gemm0N / gemm0NPerBlock;
+    int64_t gemm1NPerBlock = gemm1N;
+
+    RockAccelTuningParamAttrInterface gemm1TuningParams =
+        deriveGemm1TuningParams(rewriter, gemm0TuningParams, gemm1NPerBlock);
+    int64_t gemm1kpack = gemm1TuningParams.getKpack();
+
+    auto accelEmitterPtrGemm0 = accel::AccelEmitter::select(
+        op.getFeatures(), elemTypeQ, elemTypeK, arch, gemm0TuningParams);
+    if (!accelEmitterPtrGemm0)
+      return op.emitOpError("Unable to emit accelerator code.");
+    rock::accel::AccelEmitterParams accelParamsGemm0 =
+        accelEmitterPtrGemm0->getParams();
+    auto accelEmitterPtrGemm1 = accel::AccelEmitter::select(
+        op.getFeatures(), elemTypeV, elemTypeV, arch, gemm1TuningParams);
+    if (!accelEmitterPtrGemm1)
+      return op.emitOpError("Unable to emit accelerator code.");
+    rock::accel::AccelEmitterParams accelParamsGemm1 =
+        accelEmitterPtrGemm1->getParams();
+
+    // Get current workgroup ID.
+    auto bid = rewriter.create<WorkgroupIdOp>(loc, rewriter.getIndexType());
+    // Get current workitem ID.
+    auto tid = rewriter.create<WorkitemIdOp>(loc, rewriter.getIndexType());
+
+    // Bufers for Gemm0
+    int64_t gemm0KPerBlock = gemm0kpack * gemm0KpacksPerBlock;
+    auto [fromGlobalRegBufferQ, toLDSRegBufferQ, ldsByteBufferQ] =
+        createBuffersForGemmIn(loc, gemm0KPerBlock, blockSize, elemTypeQ,
```

Comment:
Could things like `createBuffersForGemmIn` be factored out and used in the `gridwise_gemm_accel` rewrites as well?

Reply:
It could; but not in this PR. Feel free to create a ticket if you think its a good idea and assign it to me.

Reply:
Done